### ISSUE TITLE
Revert #3002 - set type of data to any

### ISSUE THREAD

**Issue by ImRodry** [axios/axios#4141]
#### Is your feature request related to a problem? Please describe.
Yes. After #3002 was merged, we are now required to set explicit type definitions for every single axios request we make, even if we don't want to, because these would result in an error.

#### Describe the solution you'd like
The default type for AxiosResponse#data should be any so that we can modify it if we want, or keep it as is if we don't care about the typings for that specific request

#### Describe alternatives you've considered
As suggested in that PR, we could set the type to `any` explicitly, however, many people use typescript-eslint's [no-explicit-any](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-explicit-any.md) rule that prevents us from doing this. I believe everyone should be able to choose what types they wanna work with, because for extremely simple responses from which you only need one or two simple properties you don't always need explicit type declarations

#### Additional context
TypeScript's DOM library sets the return type of Body#toJSON to Promise<any>, so we should be following the native approach and not restricting user's options by setting a default that "most users may want", even if most users do not want this.

**Comment by bratanon** [comment]
+1 for this.

**Comment by tjhorner** [comment]
This seems related to #4109

**Comment by simllll** [comment]
I don't get this type at all, why is input and output the same data?

  post<T = never, R = AxiosResponse<T>>(url: string, data?: T, config?: AxiosRequestConfig<T>): Promise<R>;

post return value is Axiosreponse<T> and data to pass into is also T? Who sends the exact same data structure to a server and gets the same result? Well there are some csaes I guess, but not very common in my opinion ;-)
Input/request and output/return data should be defined seperatly.

**Comment by jasonsaayman** [comment]
Closing as the issue will be reverted in 0.23

**Comment by ImRodry** [comment]
> Closing as the issue will be reverted in 0.23

Was this a lie or what?? Reverting doesn’t mean changing the type to unknown…

**Comment by ITenthusiasm** [comment]
Is this the right place to continue the discussion @jasonsaayman ? (If not, could I be directed on what to do next, eg. open a new issue? :sweat_smile:)

I have mixed feelings about this. Personally, I'm not a big fan of the change to `unknown`.

## `unknown` is more type safe...
But I agree with @remcohaszing that it's more type safe... which is why I'm not surprised that the participants in the TS Discord discussion said to prefer `unknown` to `any`. It's just the de facto standard these days. So... there's some legitimacy there.

## ...but `any` is arguably friendlier
I think `any` is friendlier to the developer in the end. Looking at @remcohaszing's proposed question:

> Lets say you have a TypeScript function which could return anything, for example a library for making HTTP requests. The return type can be configured using a generic for convenience. Which would be the preferred default return type? any or unknown?

I wasn't part of that discussion, but I tried to skim it _briefly_ (very briefly :sweat_smile:). And unsurprisingly, I mainly saw the immediate response that I think most (and good) TS developers have: "Avoid `any` and `unknown` at all costs! We can't avoid `any`? Oh no! Well we **_have_** to use `unknown` then!" I also have this immediate response (though I'm not claiming to be a great TS dev. lol.)

In that discussion (which did not involve that many participants), I didn't see much talk about _developer experience_. And since `axios` is a library, this is something very important to consider. This discussion isn't only about whether or not our types should be "incomparably pure"; it's also about the impact on users.

When the `return type` is changed from `any` to `unknown`, it _forcibly_ shoves _all_ of the type responsibility to the developer -- even if they don't want it. "What TS dev wouldn't want types?" one may ask. Let me explain.

### Example 1
Please consider one of my repositories, which is now failing on `CI` after updating `axios`. Particularly, look at [this code](https://github.com/ITenthusiasm/ghc-podbean-helper/blob/main/server/services/PodbeanService.ts#L107-L126), which is now failing. My code attempts to interact with the `Podbean API` for podcast stuff. The API is a bit strange, the docs aren't the best, and sometimes the docs take a long time to load. PodBean exposes no "types library", so when the `return type` from `axios` is `unknown`, the responsibility is on me to create _all_ of the individual types in _my own_ repository, even though they may be subject to spontaneous change. Not only that, but I then have to cast _all_ of my interactions with `data` to these created types to do _anything_ with them in TS. In addition to being inconvenient and time consuming, this runs the risk of my formatter making one-line statements become multi-line statements, increasing LOC. (I know, a reader can consider that trivial.)

This situation gets more complicated when it comes to error handling. Caught errors, by default, are `unknown`. This means that errors need to be casted or narrowed. Life is easier if the error is known to be the instance of a certain kind of class (`instanceof` type narrowing), but I'm assuming `AxiosError` is an `interface` instead of a `class`, meaning `instanceof` narrowing is impossible. So, any time I want to interact with the `AxiosError` object, I first have to cast `err` to `AxiosError`. But after that, I have to do _more_ type casting for `data`. This quickly results in unwieldy code that becomes difficult to read. I can't even re-throw `AxiosError` or `AxiosError.response?.data` as is, because `Error`'s constructor in TS expects a string (even though it stringifies the argument).

Now sure, any individual can decide if they want to die on the hill of "supremely pure types" if they want. But once code starts getting this inconvenient, I personally feel it's unfair to _force_ this situation upon the consumer.

### Example 2
(If you're already convinced, you can skip this part. Unless you're curious for more info.)

Another example to consider is [React JSS](https://cssinjs.org/react-jss/?v=v10.8.0#dynamic-values), a library for styling React components. As you can see in the example code from the link, the library enables developers to write CSS in JS. Thankfully, it enables devs to style components _based on props_. But the problem is that the library doesn't know what the `props` will be in the defined function. Things were ultimately figured out by aggressively trying to use generic types, but let's say that, _for the sake of argument_, they had to choose _only_ between `any` and `unknown`.

`unknown` would be more "TS proper", but `any` would be _vastly_ more developer friendly. Why? These CSS declarations can get quite long, and they'll likely scale depending on the size of the component. (And not all components can be made tiny.) If the argument of the callback was typed as `unknown`, then the developer would be _forced_ (keyword) to explicitly define the type for every single callback, even though the type of `props` would already be made clear in another part of the file. This results in redundancy across several lines. If the argument of the callback was typed to `any`, then the developer has the _choice_ (keyword) of whether or not they want to endure said redundancy.

### Final "evidences", no examples
I can understand encouraging devs to use `unknown` on their on their own personal functions. But I personally don't understand the purist push for using `any` over `unknown` when _libraries_ are involved. I guess I could agree that in some cases it protects the user. And known types are _always_ preferred. But when `any` and `unknown` are literally the only options, then in many cases, I think `unknown` causes more problems. **Whether a developer is given an `any` value or an `unknown` value, they are left with the _same_ problem: The developer who wants good, safe types _has_ to explicitly type cast, type narrow, etc.** When it comes to libraries, this means that `any` satisfies the most users: 1) For the purists, they still have the option to type cast (which they'd be forced to do with `unknown` anyway), 2) For the semi-purists who like types but want code that's more readable in their eyes, they have the option to be freed from inconvenient type casting. The decision should be given to the consumer.

And as a _very_ minor tidbit, I think people like myself also like being given `any`s over `unkonwn`s from libraries (when those are the only options) because it means we can get to keep our "neat code" without having to battle TS ESLint. (The suggestion for the user to use `<any>` as a generic type argument will make lint unhappy.) That's admittedly lazy though. :sweat_smile:

Lastly, regarding that Discord conversation, the one who seemed to contribute most was `𝚠𝚎𝚋𝚜𝚝𝚛𝚊𝚗𝚍`, and I have to agree with his last statement:

> but if you have to stick with `any` for backwards compatability, I'll just suffer
it's not the end of the world

I think that's the wiser approach here. Even the native `fetch` API returns `Promise<any>` for `response.json()`. I imagine this is again for the developer experience -- as, _practically_ speaking, nothing is _gained_ for the consumer of a library when `any` is changed to `unknown`, but much is _lost_.

Open to different ideas. I know I said a lot.

**Comment by ITenthusiasm** [comment]
To touch further on that breaking changes note...

For comparison today:
`axios`: 21,505,469
`typescript`: 22,719,100

_Very_ close, with TS being slightly larger.

Even TS devs have times when they [acknowledge previous decisions that were probably mistakes](https://github.com/microsoft/TypeScript/issues/39008#issuecomment-642074983). But even then, they consider the impact of a breaking change on existing users. (Not that there are _never_ breaking changes, of course. Value, ease of fixing, how many benefit vs how many get bitten, and more come into play.)

Though related to whether or not we should go back to `any`, this is separate from the discussion on enhancing developer experiences, and is also worth considering in its own rite on the topic. The codebase I gave is small and will already require some inconvenient work if we go with `unknown`. For codebases more largely using `axios`, this breaking change could be quite large.

**Comment by remcohaszing** [comment]
The change was made from version 021.0 to 0.23.0 (The types in 0.22.0 were very broken, so let’s ignore those). As far as can tell from [semver](https://semver.org/#spec-item-4) and the [axios upgade guide](https://github.com/axios/axios/blob/master/UPGRADE_GUIDE.md), it’s ok to make breaking changes.

If the Axios team disagrees with this, I definitely think the default type should be reverted to `any`. Otherwise at I think this change should at least be documented in the migration guide.

/cc @jasonsaayman 

https://github.com/microsoft/TypeScript/issues/46347 looks very promising. That would end this discussion and similar discussions for other libraries with a compiler setting. If that get’s implemented, the correct type wouldnt be `any` nor `unknown`, but `any | unknown` of which the behaviour could be changed by the user. For now this matches `any`.

I think all arguments have been made. Ultimately the choice is up to the Axios team.

**Comment by ITenthusiasm** [comment]
> Please don’t say introducing unknown is a large inconvenience or a huge price to pay, while dismissing the added type safety as a slight convenience. These are opinions.

I don't think these are just opinions. _Preferences_ are definitely opinionated. But those aside, I think it's possible to some extent to weigh these things. Type safety is great, I agree. But the reality is that with `any`, again, there are no extra hoops that the "type purists" have to jump through. With `unknown`, _everyone else_ has many hoops to jump through. I suppose there are more ways to define inconvenience than lines of code, amount of extra research into types from API responses, extra files, lint fighting, general inertia, effort to get code working, and the other things mentioned earlier. But those are some pretty valid aspects that, again, don't seem to be addressed or compared by the opposing side yet.

I can see how "type purists" would prefer `unknown`. But no one has yet laid out major pain points. The most that has come out has been, "I'd like to get those extra errors from TS". But that's only one thing (one thing, _not_ nothing). And for people who are already type casting most of the time (key assumption), it's less of a problem (though not a completely eradicated one). I think it's fair to ask for more examples, as I did. And I think true weights can be put out and compared, albeit with bias from both sides.

One thing I'm certain of: The "summed inconvenience" between the "type purists" and everyone else on this matter is definitely unequal. People have their opinions, but inertia and mental loads are certainly measurable; and the fact that people have opinions does not change that reality. (Of course, one can argue how easily the sums can be determined. That's the point of asking for more examples.)

I'm not simply trying to assert my own preferences here. It's been about weights this whole time. Although I wouldn't like having to update my code bases, my code bases are pretty small. I could get along with using `unknown` because I typically cast types anyway. But I'm mainly concerned about about inconveniences and backwards compatibility for the consumers.

---

> Personally I even find it a bit ironic to see people who are enforcing this rule in their own code base aren’t happy with the change to unknown, because this seems to align exactly with the goal of that rule (to avoid any).

There can be a few reasons. For one thing, some "semi type purists" likely have a mental distinction between the types they have control over, and the types they don't have control over. And for unavoidable jank/weird APIs, it can also be difficult to know how to organize and name types at times. There's also the chance that others arguing against `unknown` share concerns for general users.

There are additional reasons people could be dissatisfied.

---

> I expected this to be the main reason for using `any` in the `dom` types, which apparently isn’t true.

I'm not surprised that `any` was used. And I'm not surprised that it wasn't changed to `unknown` when that type came out. Other things have been changed to `unknown` over time, like the `err` in a `catch` block. I think it boils down to minimizing pain points for the overall user base, as orta mentioned. (And the fact that he mentioned it, again, shows that inconvenience is measurable to some extent. It's not just opinions.)

---

> 3. Create a poll

Possible. I personally don't think a poll is necessary to resolve this discussion. But if a poll _was_ used, I think that for the sake of comparison, `axios` should be put alongside things like `JSON.parse()` and `fetch`'s `response.json()`. This would help add clarity and get at the heart of the matter.

---

Regarding the 4th option, I've also been trying to think of non-hackish workarounds (for `axios` to the users, not for the users), though I haven't had enough time. Nothing special has come to mind yet.

**Comment by ITenthusiasm** [comment]
Is it at all possible to generate predictable types for function calls based on key-value pairs?

Typically, well-documented APIs will say, "If you METHOD URL, you get back TYPE". Theoretically, the METHOD-URL-TYPE relationship could be represented in an `interface`.

Would it be possible to pass such an interface somewhere so that axios would know the proper types without type casting? :thinking: (I'm sure _if_ it's possible, that would be effort. But the question remains. :eyes:)

**Comment by ITenthusiasm** [comment]
@remcohaszing appreciate your patience!!!

**Comment by remcohaszing** [comment]
The main reason I believe a poll could help, is because ultimately it’s up to the Axios team to choose based on a heated discussion. As far as I could tell so far the Axios team isn’t really involved in TypeScript, which makes it difficult for them to make a decision.

---

As for option 4, this is what it could look like:

```ts
interface AxiosTypes {
  // defaultResponseData: any;
}

type AxiosResponseData = AxiosTypes extends { defaultResponseData: infer T } ? T : unknown;

declare function request<T = AxiosResponseData>(): { data: T };

const { data: explicitType } = request()

const { data: implicitType } = request<number>()
```

This is what end user code would look like:

```ts
declare module 'axios' {
  interface AxiosTypes {
    defaultResponseData: any;
  }
}
```

Here’s a [playground link](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgIIA9gHsDOAVATwAcIdkBvAKGWQHpbkATCeAVwBswAlUorEHBAAicMHABcyOCAIBuSgF9KlMMRQZsOHjj4DhouMgC8aTLkIkyEdJBCMy5JizgduvfoJFjJoGNGR4yArIAPwByJKsIADWIFgA7iDylMwI7HBQKDBRCGDYIMiZAI6spGAAPIEmGrjaup4GAHwAFACUko6MBpKBCskIHmAUTN3IwAC2ROzACMBgFijBJsWlOGBtygMCQ52j1lMzcwtBxoUQJWXlIKzjAEbQLa2UQA). You can see the types of `implicitType` and `explicitType` by hovering over them. The potential interface augmentation can be done by toggling the commented line, or you could use whatever default type you like :shrug:. In this case the default defaults to `unknown`, but it could also default to `any`.

It’s a bit hacky, and a library specific solution, but at least this settles everyone’s preferences.

**Comment by ITenthusiasm** [comment]
> Suggesting a user to change a library's typings in order to fix an issue it introduced is never a good thing

That depends on your definition of "fix", though. Corrected types in libraries will oftentimes be a breaking change, but they are never "problems" or "errors" on the maintainer's part, per se. There's room for disagreement on whether or not the change from `any` to `unknown` is a correction, but I don't think the argument can be made that it's a "problem" or "error".

Option 4 was just a proposition as a middle ground. If the Axios team determined `unknown` to be the correct type, then option 4 basically provides an easy migration. But it wouldn't be a hack to fix a problem that's the responsibility of the maintainers.

> Having the type default to `any` makes both parties happy

I think the discussion has shown that not everyone would be fully satisfied with `any`. :sweat_smile: That's what makes the views from both sides important. But I think all the views have been expressed. Aside from my retracting some invalid points, I think the topic is approaching circularity.

**Comment by jekh** [comment]
> > I think more creative people could come up with an even better solution for eslint no-explicit-any users who prefer the former implicit any return type.
> 
> Not sure about that one.

Just wanted to point out that @remcohaszing did indeed come up with a much more elegant solution in Option 4 than the workaround I proposed :)

**Comment by jasonsaayman** [comment]
I will review this tonight and make the needed decision :)

**Comment by bobvandevijver** [comment]
Not to spark any discussion whatsoever, but I just wanted to let you guys know that I actually loved the change to `unknown` as that forced me to correct the typings on many places, resulting in better typed code. 

I would love to see the `unknown` come back in a v1 release in the future (I already did the work anyways 😅, but I can understand that this change does not fit within the current v0 release. Even though you could discuss that a v0 is not stable (so changes like these can just "happen"), that is a whole other discussion 😄

**Comment by ITenthusiasm** [comment]
Yeah some kind of RFC leading up to v1 could be useful. (Or the discussion here could be re-used when v1 comes along too, I suppose. :thinking:) If microsoft/TypeScript#46347 gets addressed in time, it may not even need to be a big discussion.

**Comment by paulnelson2** [comment]
So it's not a sure thing that v1 will have unknown?

In that case, let me give my two cents:

I think unknown would be the default for data, and with all due respect to those discussing whether unknown or any is better in general, that's not the point. The point I think can be shown through a comparison:

1. Default unknown -- In this situation, those who prefer unknown have no work, and those who prefer any will be told by Typescript in each and every place that they need to change to any, and then they can do so.
2. Default any -- In this situation, those who prefer unknown will never be informed by Typescript where they need to change things. Naturally, those who prefer any have no work to do.

In summary, (1) results in visible problems for some people, and (2) results in invisible problems for some people. Some people are going to have problems either way. I would argue that causing users visible problems is better than causing users invisible problems and that therefore (1) default unknown is the best choice.

That being said, I'm happy to wait until v1 cuz breaking changes.

------------------------------

@KosnIre 

> I see the suggestion:
> 
> > It’s easy to specify the response data should be any.
> 
> ```ts
> const { data } = axios.get<any>('/');
> // data is now any
> ```
> 
> So for those who want the type benefits of `unknown`, wouldn't the equivalent be possible?
> 
> ```ts
> const { data } = axios.get<unknown>('/');
> // data is now unknown
> ```
> 

The problem is that those who want to do `axios.get<any>` will be told by Typescript where they need to make changes, but the people who would want `axios.get<unknown>` won't be told.

**Comment by cgarrovillo** [comment]
TypeScript is only as good as the developer that uses it. Per the official [TypeScript docs](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#any) regarding the use of `any`:

> any
❌ Don’t use any as a type unless you are in the process of migrating a JavaScript project to TypeScript. The compiler effectively treats any as “please turn off type checking for this thing”. It is similar to putting an @ts-ignore comment around every usage of the variable. This can be very helpful when you are first migrating a JavaScript project to TypeScript as you can set the type for stuff you haven’t migrated yet as any, but in a full TypeScript project you are disabling type checking for any parts of your program that use it.

> In cases where you don’t know what type you want to accept, or when you want to accept anything because you will be blindly passing it through without interacting with it, you can use unknown.

And from the beginning of the thread: 
> because for extremely simple responses from which you only need one or two simple properties you don't always need explicit type declarations

If it is so simple, why not just declare the types?  That just sounds lazy and incompetent.
**What is the point of using the TypeScript language at all if you're choosing to deliberately go back to JS using `any`?** I am all for developer experience, and I get `any` will be the easier route always. 

But on the same coin, failing to provide types just results in technical debt, a future debt repaid when a new developer comes on to the project and works with the axios logic written by someone who didn't want to take a few minutes to declare the types of "extremely simple responses".

**Comment by popuguytheparrot** [comment]
Types not fixed yet 

```
   59 | RequestHttp.interceptors.request.use((config) => {
    60 |     const token = JwtTokenLocalStorageService.getParsedAccessToken();
  > 61 |     config.headers.Authorization = token ? `Bearer ${token}` : '';
       |     ^
```

**Comment by ImRodry** [comment]
@popuguytheparrot that issue is completely unrelated to the original issue reported here
